### DISCUSSION TITLE
How to use useQuery inside the useEffect

### DISCUSSION THREAD

**Discussion by 989x** [TanStack/query#4873] [Q&A] [answered]
By default, useEffect values are automatically changed when name in ```products?search={name}``` has changed.

But in tanstack useQuery there is a problem that, useQuery cannot be used in useEffect.

### Sample case
1. Before query  
http://localhost:3000/product?search=iphone13
return iphone13 data

2. After query
http://localhost:3000/product?search=iphone14
return iphone13 data 

But I hope to return iphone14 data automatically.

### Sample code

Now doing,  But it works only when refreshing the web.
```ts
    const [posts, setPosts] = useState([]);

    const { isLoading, error, data } = useQuery({
        queryKey: ['products'],
        queryFn: () =>
        productFind({name: props.name}),
    })
    console.log("isLoading", data)

    if(isLoading) return <div>loading...</div>

    useEffect(() => {
        setPosts(data?.productFind), [data?.productFind]
    })
```

Hope to see like this
```ts
    const [posts, setPosts] = useState([]);

    useEffect(() => {
        const { isLoading, error, data } = useQuery({
            queryKey: ['products'],
            queryFn: () =>
            productFind({name: props.name}),
        })
        console.log("isLoading", data)

        setPosts(data?.productFind), [data?.productFind]
    })
```

But found error
```
Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component. 
This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
```

**Comment by louis-young** [comment, marked as answer]
Hi 👋 

Firstly, hooks can only be called at the top level of a function component or hook, not within an effect callback function.

To achieve what you're looking for, you just need to include the name of the product you want to search for in the query key. This will mean that when the product name is updated, a new query is created that will run the query function for the new product name.

Please also see: https://tkdodo.eu/blog/effective-react-query-keys

As an aside, you probably don't need to derive the `posts` state from the data returned from the query hook. Once you start calling the hook at the top level and using the suggested query key, `data` will contain whatever the query function successfully resolves with and be in the necessary scope.

Hopefully this helps 🙂

**Comment by 989x** [comment]
Thank you a lot for your answer @louis-young 🙏🏻

and I've found a solution, follow the instructions of the following question post.

1. #4844 

2. #2347

### Solution

My code can solve the problem by just adding ```props.name``` into the ```queryKey``` and after that useQuery can be used in useEffect.

```ts
    const [posts, setPosts] = useState([]);

    const { isLoading, error, data } = useQuery({
        queryKey: ['products', props.name],
        queryFn: () =>
        productFind({name: props.name}),
    })
    console.log("isLoading", data)

    if(isLoading) return <div>loading...</div>

    useEffect(() => {
        setPosts(data?.productFind), [data?.productFind]
    })
```

That's all, useQuery alaways change when query client has change.

**Reply by louis-young** [reply]
No worries 🙂 

For what it's worth if you want to derive another piece of state from the data returned by the query (which you usually don't want to do, for the record) you can do this in an `onSuccess` callback and omit the effect entirely.

Tangentially, your effect probably isn't doing what you think it is. The first argument to `useEffect` is a callback function and the second argument is the dependency array. You're only passing a single argument and it looks like you might be trying to supply a dependency array.

I'm glad I could help, have fun!
